### DISCUSSION TITLE
BBC Sounds Radio Player

### DISCUSSION THREAD

**Discussion by joshdavenport** [johnlindquist/kit#913] [Share]
![2022-10-15 16 00 35](https://user-images.githubusercontent.com/757828/195993376-ac8fe206-1903-4c34-8a50-b996d6cae402.gif)

[Open bbc-sounds in Script Kit](https://scriptkit.com/api/new?name=bbc-sounds&url=https://gist.githubusercontent.com/joshdavenport/5d6b0f3207225fe9091ee0573daa177d/raw/cea82f55c39ebfd51e64d966871f3c3a99304534/bbc-sounds.ts")

Play the main BBC Radio stations, with a nice little visual as to what the current show is. Easy to add new station just by changing the choices passed to `station`. Forgive me for putting 6music first, although, I'm not really sorry about that 😛 

I'm assuming this won't work well for anybody outside of the UK unfortunately.

```js
// Name: BBC Sounds Player
// Description: Stream a radio station from BBC Sounds
// Author: Josh Davenport-Smith
// Twitter: @jdprts

import "@johnlindquist/kit";
import { WidgetAPI } from "@johnlindquist/kit/types/pro";

const PLAYER_WIDTH = 180;
const PLAYER_HEIGHT = 195;

const getStyles = () => `
:root {
    --player-width: ${PLAYER_WIDTH}px;
    --player-height: ${PLAYER_HEIGHT}px;
    --header-offset: 25px;
    --footer-offset: 25px;
    --scale: 2.2;
}
 
#player {
    width: var(--player-width);
    height: var(--player-height);
    overflow: hidden;
    position: relative;
}
 
iframe {
    position: absolute;
    left: 0;
    width: calc(var(--player-width) * var(--scale));
    height: calc((var(--player-height) * var(--scale)) + (var(--header-offset) * var(--scale)) + (var(--footer-offset) * var(--scale)));
    top: calc(var(--header-offset) * -1);
    transform: scale(calc(1 / var(--scale)));
    transform-origin: 0 0;
}
`;

const getScripts = (station: string) => `
(() => {
    const playerContainer = document.getElementById('player');
    const url = 'https://www.bbc.co.uk/sounds/player/bbc_${station}';
    const iframe = document.createElement('iframe');
    iframe.setAttribute('src', url);
    iframe.setAttribute('enablejsapi', true);
    playerContainer.appendChild(iframe);
})();
`;

const station = await arg('Select station', [
    {
        name: '6 Music',
        value: '6music',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_6music/colour_default.svg',
    },
    {
        name: 'Radio 1',
        value: 'radio_one',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_radio_one/colour_default.svg',
    },
    {
        name: 'Radio 2',
        value: 'radio_two',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_radio_two/colour_default.svg',
    },
    {
        name: 'Radio 3',
        value: 'radio_three',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_radio_three/colour_default.svg',
    },
    {
        name: 'Radio 4',
        value: 'radio_fourfm',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_radio_four/colour_default.svg',
    },
    {
        name: 'Radio Five Live',
        value: 'radio_five_live',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_radio_five_live/colour_default.svg',
    },
    {
        name: 'World Service',
        value: 'world_service',
        img: 'https://sounds.files.bbci.co.uk/3.1.5/networks/bbc_world_service/colour_default.svg',
    },
])

const w: WidgetAPI = await widget(
    `
<div id="player"></div>
<style type="text/css">${getStyles()}</style>
<script type="text/javascript">${getScripts(station)}</script>
    `,
    { alwaysOnTop: true }
);


w.setSize(PLAYER_WIDTH, PLAYER_HEIGHT);

w.onClick(event => event.targetId === "x" && w.close());

w.onResized(() => w.setSize(PLAYER_WIDTH, PLAYER_HEIGHT));
```

**Comment by johnlindquist** [comment]
Love it! Fyi, you can toss the width and height in the widget options and avoid that "onResize" handler.

Widget Options use all of the Browser window constructor options:

https://www.electronjs.org/docs/latest/api/browser-window#class-browserwindow

**Comment by johnoscott** [comment]
OMG so cool (works for me in Australia) I found one weird issue though with moving the window on MacOS (Monterey) . When i click-and-drag on the logo, said logo appears bigger and transparent.
![screenshot-20221021- johnlindquistkit  BBC Sounds Radio Player (Discussion #913) - johnoscott@gmail com - Gmail-001288 png](https://user-images.githubusercontent.com/291958/197087759-59581a07-3b0f-4d6e-8105-2d8f0da42f88.gif)

**Reply by joshdavenport** [reply]
Hah, this certainly is an odd one. Probably some combination of the scaling that happens with the iframe and some internal Chromium/Electron oddity. I reckon it's probably best to add a nice draggable area like with my other widget https://github.com/johnlindquist/kit/discussions/920. Will take a look

**Comment by johnoscott** [comment]
Can I also suggest some additional features :
* close box
* switch to different channel

**Reply by joshdavenport** [reply]
Fair suggestion with the close button, main reason I didn't was because I've just been using Cmd + w, so no X button was just a bit nicer looking. Re switching, I wonder if it's possible to, if the widget is closed, re-present the `arg` station selector.

**Reply by johnlindquist** [reply]
@joshdavenport Here's a solution to "re-present" the `arg`:

```js
// Name: Testing Reopen Widget

import "@johnlindquist/kit"

let createWidget = async () => { 
    let message = await arg("Message")
    let w = await widget(`<div class="text-4xl">${message}</div>`)
    w.onClose(createWidget)
}

createWidget()
```

[Open testing-reopen-widget in Script Kit](https://scriptkit.com/api/new?name=testing-reopen-widget&url=https://gist.githubusercontent.com/johnlindquist/72879e550b3828daf47598fce3c44037/raw/28f70187c2f0fda12170fb1c1ff1b96fb6533746/testing-reopen-widget.ts")

**Reply by johnlindquist** [reply]
@johnoscott @joshdavenport 

Also, here's a demo of "display close buttons on hover":
```js
// Name: Widget with Close Buttons

import "@johnlindquist/kit"

await widget(`Hover the top left to expose buttons`, {
    titleBarStyle: 'customButtonsOnHover',
})
```

[Open widget-with-close-buttons in Script Kit](https://scriptkit.com/api/new?name=widget-with-close-buttons&url=https://gist.githubusercontent.com/johnlindquist/d8e25b9cd063e5603d0f62b29472cbab/raw/7521228211007b167a7d7b282a3b18a362a55784/widget-with-close-buttons.ts")

**Reply by johnlindquist** [reply]
In fact, I'm going to make the `titleBarStyle: 'customButtonsOnHover',` the default for widgets since they only show up when you hover and it feels like an expected behavior. We can still disable it with `titleBarStyle: 'hidden',`

**Reply by joshdavenport** [reply]
That sounds completely reasonable. Thanks for your previous suggestions too, I've got some changes in the work for this widget. I'm mainly working out how best to deal with the player wanting you to sign in. Using an iframe the sign in page breaks due to the value of `X-Frame-Options` and using a webview where I can set `disablewebsecurity` makes the link open in default chrome, and twice. Experimenting with injecting some scripts into the page to change behaviour but haven't explored that too much yet.
